process_settings.go

```
@@ -0,0 +1,200 @@
+package process_settings
+
+import (
+	"errors"
+	"fmt"
+	"log"
+	"os"
+	"reflect"
+	"strings"
+
+	"github.com/fsnotify/fsnotify"
+	"gopkg.in/yaml.v3"
+)
+
+// A ProcessSettings is a collection of settings files and a target evaluator
+// that can be used to get the value of a settings based on the current targeting.
+type ProcessSettings struct {
+	FilePath            string            // The path to the settings file that was used to create the ProcessSettings
+	Settings            *[]SettingsFile   // The settings files that make up the ProcessSettings
+	TargetEvaluator     TargetEvaluator   // The target evaluator that is used to determine which settings files are applicable
+	Monitor             *fsnotify.Watcher // The file monitor that is used to detect changes to the settings file
+	WhenUpdatedRegistry []func()          // A list of functions to call when the settings are updated
+}
+
+// NewProcessSettingsFromFile creates a new instance of ProcessSettings by
+// loading the settings from a specified file path and using the specified
+// static context to evaluate the targeting.
+func NewProcessSettingsFromFile(filePath string, staticContext map[string]interface{}) (*ProcessSettings, error) {
+	settings, err := loadSettingsFromFile(filePath)
+	if err != nil {
+		return nil, err
+	}
+
+	monitor, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, err
+	}
+
+	err = monitor.Add(filePath)
+	if err != nil {
+		return nil, err
+	}
+
+	return &ProcessSettings{
+		FilePath:        filePath,
+		Settings:        settings,
+		TargetEvaluator: TargetEvaluator{staticContext},
+		Monitor:         monitor,
+	}, nil
+}
+
+// Get returns the value of a setting based on the current targeting.
+// If the requested setting is not found, an error is returned.
+func (ps *ProcessSettings) Get(settingPath ...interface{}) (interface{}, error) {
+	value := ps.SafeGet(settingPath...)
+
+	if value == nil {
```

Because settings are JSON, they're allowed to have `null` values, which I think will come back as `nil` here. Can you flip these and have `SafeGet` call `Get` and suppress "The setting '...' was not found error?

---

I'm wondering if there is even a need for `SafeGet` in golang. Do we need it if using `SafeGet` is the same as doing:
```go
value, _ := process_settings.Get(...)
```

I'm leaning toward simplifying the contract given error handling patterns in golang. Thoughts?

---

I don't think that's actually the same. Suppose we had an exception because we couldn't parse the JSON file. We wouldn't want to treat that if the setting were missing. That would be equivalent to the Ruby anti-pattern of `rescue nil`.

If you have offer a `SafeGet`, that could intercept the exact exception where it's missing right?

(Uggh, it's ridiculous in 2023 to have to think about hand-writing exception contracts.)

---

Agreed. But also the exceptions coming from failure to parse will come from the `fsnotify` watcher goroutine and the instantiating of the ProcessSettings itself. The former of which would swallow the exception and log out an error, and the latter returns an error to be acted on independently. Additionally in the goroutine, we don’t swap out the current settings if we get a parsing error, so that won’t leave us will bad settings data. 

The error here is simply that a setting doesn’t exist or is explicitly set to `nil` when being requested. Which can be swallowed if the developer doesn’t care about settings being nil. 

All this to say, I’m still not sure this is necessary because parse errors happen elsewhere to be handled, and `Get` will only ever error on a nil value being returned which allows for ignoring the error to have `SafeGet`. 

But I [also see](https://github.com/Invoca/process_settings/blob/master/lib/process_settings/abstract_monitor.rb#L127) that we allow for `nil` to be explicitly set which is not currently supported in the `Get` case which is likely more the problem here that I need to fix. Currently if we set a setting explicitly to `nil` it will return an error in the `Get` interface. I’ll get that sured up and shake things loose a bit more here to see if this `SafeGet` is still necessary.

---

Got the functionality put in to support explicitly set `nil` process settings with 372c2d5070ce9b8ba6aaa08e2bda4ce1c28ac60b. That refactor made it a lot more evident that we should have `SafeGet` call `Get` since `Get` needs to discern between `nil` and `not_set` values. Still have the question regarding the actual need for `SafeGet` since it's just swallowing any errors from `Get` the same way caller code could, but that could be debated later for sure.